### PR TITLE
Lint rule to detect unminified errors

### PR DESCRIPTION
**Pull request by acdlite**
## Based on #22428

Adds a lint rule that detects when an Error constructor is used without a corresponding production error code.

We already have this for `invariant`, but not for regular errors, i.e. `throw new Error(msg)`. There's also nothing that enforces the use of `invariant` besides convention.

There are some packages where we don't care to minify errors. These are packages that run in environments where bundle size is not a concern, like react-pg. I added an override in the ESLint config to ignore these.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -127,6 +127,41 @@ module.exports = {
   },
 
   overrides: [
+    {
+      // By default, anything error message that appears the packages directory
```
Comment (rickhanlonii) [inline .eslintrc.js:131]:
```suggestion
      // By default, any error message that appears the packages directory
```

File: packages/react-client/src/ReactFlightClient.js [hunk 2]
```diff
@@ -397,6 +397,7 @@ export function resolveError(
   message: string,
   stack: string,
 ): void {
+  // eslint-disable-next-line react-internal/prod-error-codes
```
Comment (rickhanlonii) [inline packages/react-client/src/ReactFlightClient.js:400]: Maybe include a reason why?

Comment (rickhanlonii) [reply packages/react-client/src/ReactFlightClient.js:400]: Ditto elsewhere

File: packages/react-client/src/ReactFlightClientHostConfigNoStream.js [hunk 3]
```diff
@@ -12,19 +12,22 @@ export type StringDecoder = void;
 export const supportsBinaryStreams = false;
 
 export function createStringDecoder(): void {
+  // eslint-disable-next-line react-internal/prod-error-codes
```
Comment (rickhanlonii) [inline packages/react-client/src/ReactFlightClientHostConfigNoStream.js:15]: What about defining something like an `NotImplementedError` or `UnminifiedError` which is automatically ignored by the lint rule and parser?

**Review by rickhanlonii — APPROVED** [review]
Not blocking, but I feel like needing to disable this many instances is a smell. Why are so many disabled?

**Comment by acdlite** [comment]
> Not blocking, but I feel like needing to disable this many instances is a smell. Why are so many disabled?

It reflects that we're not super consistent about which errors we minify and which ones we don't, because there's no existing lint rule that enforces it either way.

The first step should be to decide which packages should have minified error messages. I think for browser builds, probably everything.

If there are exceptions after that we can consider adding an escape hatch, though I would still prefer an inline comment directive (like `// no-minify-error-message` or something) since it doesn't have runtime implications and we can adapt it to work with any Error subclass, like TypeError for example.

**Pull request closed without merging**